pyteal/ast/subroutine.py

```diff
@@ -810,53 +941,146 @@ def var_n_loaded(
+            )
+            internal_abi_var._stored_value = FrameVar(proto, dig_index)
+            argument_var = None
+            loaded_var = internal_abi_var
+        else:
+            dig_index = (
+                subroutine.arguments().index(param) - subroutine.argument_count()
+            )
+            argument_var = None
+            loaded_var = FrameVar(proto, dig_index).load()
 
-    arg_vars: list[ScratchVar] = []
-    loaded_args: list[ScratchVar | Expr | abi.BaseType] = []
-    for arg in args:
-        arg_var, loaded_arg = var_n_loaded(arg)
-        arg_vars.append(arg_var)
-        loaded_args.append(loaded_arg)
+        return argument_var, loaded_var
 
-    abi_output_kwargs: dict[str, abi.BaseType] = {}
-    output_kwarg_info = OutputKwArgInfo.from_dict(subroutine.output_kwarg)
-    output_carrying_abi: Optional[abi.BaseType] = None
+    @staticmethod
+    def __proto(subroutine: SubroutineDefinition) -> Proto:
+        arg_stack_types: list[TealType] = []
+        for t in subroutine.expected_arg_types:
+            if t is Expr:
+                arg_stack_types.append(TealType.anytype)
+            elif t is ScratchVar:
+                arg_stack_types.append(TealType.uint64)
+            else:
+                arg_stack_types.append(cast(abi.TypeSpec, t).storage_type())
 
-    if output_kwarg_info:
-        output_carrying_abi = output_kwarg_info.abi_type.new_instance()
-        abi_output_kwargs[output_kwarg_info.name] = output_carrying_abi
+        num_return_allocs: int = int(subroutine.has_abi_output)
 
-    # Arg usage "B" supplied to build an AST from the user-defined PyTEAL function:
-    subroutine_body = subroutine.implementation(*loaded_args, **abi_output_kwargs)
+        local_stack_types: list[TealType] = []
+        if subroutine.has_abi_output:
+            output_info = cast(
+                OutputKwArgInfo, OutputKwArgInfo.from_dict(subroutine.output_kwarg)
+            )
+            local_stack_types = [output_info.abi_type.storage_type()]
 
-    if not isinstance(subroutine_body, Expr):
-        raise TealInputError(
-            f"Subroutine function does not return a PyTeal expression. Got type {type(subroutine_body)}."
+        layout: ProtoStackLayout = ProtoStackLayout(
+            arg_stack_types, local_stack_types, num_return_allocs
+        )
+
+        # if subroutine do not have abi output, then only two cases happen:
+        # - subroutine is a normal subroutine, then check subroutine body evaluates to something, rather than none
+        # - subroutine is an ABIReturnSubroutine, the type is void, and its subroutine body type of is always none
+        num_stack_outputs: int = (
+            1
+            if subroutine.has_abi_output
+            else int(subroutine.return_type != TealType.none)
         )
 
-    deferred_expr: Optional[Expr] = None
+        return Proto(subroutine.argument_count(), num_stack_outputs, mem_layout=layout)
```

Comment:
Following up from https://github.com/algorand/pyteal/pull/584#discussion_r1022155872, it appears that there _is_ a case where `num_stack_outputs != num_return_allocs`, and that is when there's a non-ABI subroutine that returns a value. Because the return value of such a subroutine is the last thing it places on the stack, it's not easy to make it use a preallocated stack return value in any interesting way.

Note that we have to solve this problem if we want to use non-return local variables in a subroutine like that, since just placing the return value on top of the stack will no longer be good enough. But that's a problem for another day.